runtime/dynstats.c

```diff
@@ -624,3 +947,165 @@ dynstats_inc(dynstats_bucket_t *b, uchar* metric) {
 	RETiRet;
 }
 
+static struct wrkrInfo_s {
+	pthread_t tid;	/* the worker's thread ID */
+} wrkrInfo;
+
+typedef struct file_write_entry_s {
+	STAILQ_ENTRY(file_write_entry_s) link;
+	const char* name;
+	const char* content;
+	dynstats_bucket_t *bucket;
+} file_write_entry_t;
+
+typedef struct file_write_queue_s {
+	STAILQ_HEAD(head, file_write_entry_s) q;
+	STATSCOUNTER_DEF(ctrEnq, mutCtrEnq);
+	int size;
+	int ctrMaxSz;
+	statsobj_t *stats;
+	pthread_mutex_t mut;
+	pthread_cond_t wakeup_worker;
+} file_write_queue_t;
+
+static file_write_queue_t work_q;
+static int wrkrRunning;
+
+
+static void
+processWorkItem(file_write_entry_t *task) {
+	// write out the file
+	DBGPRINTF("dynstats: worker processing - '%s', '%s'\n", task->name, task->content);
+	writeStateFile(task->bucket, task->name, task->content);
+}
+
+static void *
+fileWriteWorker(void* data) {
+	struct wrkrInfo_s *me = (struct wrkrInfo_s*) data;
+	pthread_mutex_lock(&work_q.mut);
+	++wrkrRunning;
+	pthread_mutex_unlock(&work_q.mut);
+
+	file_write_entry_t *task;
+	while(1) {
+		task = NULL;
+		pthread_mutex_lock(&work_q.mut);
+		if (work_q.size == 0) {
+			--wrkrRunning;
+			//if (glbl.GetGlobalInputTermState() != 0) {
+			// TODO: clean this up
+			if(ATOMIC_FETCH_32BIT(&bTerminateInputs, &mutTerminateInputs) != 0) {
```

Comment:
we are also directly calling `destroyFileWriteWorker` when `dynstats_destroyAllBuckets` is called, but agree, we can just add `stopFlush` instead of using the global shutdown state.